### PR TITLE
[aes] Add signature conversion decorator to `aes`

### PR DESCRIPTION
**Pull request by sulyi**
### Before submitting a *pull request* make sure you have:
- [ ] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [ ] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

It starts an overhaul of `aes`.
- [x] Simplifies signature of functions in the module, instead list of integers they accept bytes.
- [x] Adds function for AES-128-GCM  decrypt (uses already existing function for AES-CRT) and verification.
- [ ] Adds new functionality to existing code (i.e. AES-CRT nonce)
- [ ] Converts (internal) functions to also use mentioned types.
- [ ] Restructures the code according to OOP design
- [ ] Adds new tests cases to `test_aes` to safeguard code

I'd like to mention @pukkandan and the help he gave me during  the development of this code at https://github.com/yt-dlp/yt-dlp/pull/938.

### CODE REVIEW

**Comment by sulyi** [comment]
I have the test just haven't added them yet.

**Comment by dirkf** [comment]
Why have the `aes_...` functions been given a leading `_`, contrary to PEP8?

Is this halfway to refactoring the code with these functions as static methods of some class?

The list `__all__` of functions in the module `aes.py` doesn't have the `_`.

**Comment by sulyi** [comment]
> Why have the aes_... functions been given a leading _, contrary to PEP8?

There's no contradiction here, they are intended yo be private.

> Is this halfway to refactoring the code with these functions as static methods of some class?

I wouldn't use static but, kind of yes. Moving functions into a class is secondary, but rewriting them to use byte instead of list of ints is.

> The list __all__ of functions in the module aes.py doesn't have the _.

Private functions are exposed [here](https://github.com/sulyi/youtube-dl-rg3_epitaph/blob/aa6f0402766f74916a35932764997812c9568cbc/youtube_dl/aes.py#L232-L264).

File: youtube_dl/aes.py [hunk 1]
```diff
@@ -80,50 +111,75 @@ def aes_cbc_encrypt(data, key, iv):
...
         previous_cipher_block = encrypted_block
 
     return encrypted_data
 
 
-def key_expansion(data):
+def _aes_gcm_decrypt_and_verify(data, key, tag, nonce):
     """
-    Generate key schedule
+    Decrypt with aes in GBM mode and checks authenticity using tag
```
Comment (pukkandan) [inline youtube_dl/aes.py:124]: Since #29201 is not merged, youtube-dl does not have any need for gcm

**Comment by dirkf** [comment]
OK, now I see the new block (which after all is mentioned in the title of the PR!) that all makes sense, thanks.

Making the lists of int internal certainly seems a good idea. The decorator mechanism is slick and functional, and systematises the code, even if it doesn't always reduce it by much. Eg AIUI these are equivalent?
```
@signature_conversion(convert_all_args_to_intlist, None, intlist_to_bytes)
def aes_ctr_decrypt(data, key, iv):
    return _aes_ctr_decrypt(data, key, iv)
```
vs
```
def aes_ctr_decrypt(data, key, iv):
    return intlist_to_bytes(
        _aes_ctr_decrypt(bytes_to_intlist(data), bytes_to_intlist(key), bytes_to_intlist(iv)))
```
But some of the other interface signatures need more complex munging and so should be more simply expressed by the decorator.

**Comment by sulyi** [comment]
Nope, that's is done [here](https://github.com/sulyi/youtube-dl-rg3_epitaph/blob/aa6f0402766f74916a35932764997812c9568cbc/youtube_dl/aes.py#L14-L15) and [here](https://github.com/sulyi/youtube-dl-rg3_epitaph/blob/aa6f0402766f74916a35932764997812c9568cbc/youtube_dl/aes.py#L21-L22).

But, @dirkf, you got me thinking whether it is possible to slap decorator directly on original functions.

File: youtube_dl/aes.py (hunk 1 above)
Comment (sulyi) [reply youtube_dl/aes.py:124]: True, since I did it I just want it out somewhere...

**Comment by pukkandan** [comment]
> you got me thinking whether it is possible to slap decorator directly on original functions.

If you never need access to the original functions (even internally), yes. You can just put the decorator on the original without needing a wrapper

**Comment by sulyi** [comment]
My original assumption that using `bytes` would be better is a bust: https://bugs.python.org/issue19251.
But I was yet able to make some improvements, and imho, the new signature is still friendlier.

**Pull request closed without merging**